cpp/ql/src/Likely Bugs/OO/UnsafeUseOfThis.ql

```diff
@@ -0,0 +1,204 @@
+) {
+  not f.isVirtual() and
+  call.getStaticCallTarget() = f and
+  init.getEnclosingFunction() = f and
+  call.getThisArgument() = instr and
+  init.getIRVariable() instanceof IRThisVariable
+}
+
+/** Holds if `instr` is a `this` pointer of type `c` used by the call instruction `call`. */
+predicate isSink(Instruction instr, CallInstruction call, Class c) {
+  exists(PureVirtualFunction func |
+    call.getStaticCallTarget() = func and
+    call.getThisArgument() = instr and
+    instr.getResultType().stripType() = c and
+    // Weed out implicit calls to destructors of a base class
+    not func instanceof Destructor
+  )
+}
+
+/** Holds if `init` initializes the `this` pointer in class `c`. */
+predicate isSource(InitializeParameterInstruction init, string msg, Class c) {
+  (
+    exists(Constructor func |
+      not func instanceof CopyConstructor and
+      not func instanceof MoveConstructor and
+      func = init.getEnclosingFunction() and
+      msg = "construction"
+    )
+    or
+    init.getEnclosingFunction() instanceof Destructor and msg = "destruction"
+  ) and
+  init.getIRVariable() instanceof IRThisVariable and
+  init.getEnclosingFunction().getDeclaringType() = c
+}
+
+/** Holds if `instr` flows to a sink. */
+predicate flowsToSink(Instruction instr) {
+  isSink(instr, _, _)
+  or
+  exists(Instruction mid |
+    successor(instr, mid) and
+    flowsToSink(mid)
+  )
+}
+
+/** Holds if `instr` flows from a source. */
+predicate flowsFromSource(Instruction instr) {
+  isSource(instr, _, _)
+  or
+  exists(Instruction mid |
+    successor(mid, instr) and
+    flowsFromSource(mid)
+  )
+}
+
+/**
+ * Holds if `instr` is an argument (or argument indirection) to a call, and
+ * `succ` is the corresponding initialization instruction in the call target.
+ */
+predicate flowThroughCallable(Instruction instr, Instruction succ) {
+  flowsToSink(succ) and
+  (
+    // Flow from an argument to a parameter
+    exists(CallInstruction call, InitializeParameterInstruction init | init = succ |
+      getPositionalArgumentInitParam(call, instr, init, call.getStaticCallTarget())
+      or
+      getThisArgumentInitParam(call, instr, init, call.getStaticCallTarget())
+    )
+    or
+    // Flow from argument indirection to parameter indirection
+    exists(
+      CallInstruction call, ReadSideEffectInstruction read, InitializeIndirectionInstruction init
+    |
+      init = succ and
+      read.getPrimaryInstruction() = call and
+      init.getEnclosingFunction() = call.getStaticCallTarget()
+    |
+      exists(int n |
+        read.getSideEffectOperand().getAnyDef() = instr and
+        read.getIndex() = n and
+        init.getParameter().getIndex() = unbind(n)
+      )
+      or
+      call.getThisArgument() = instr and
+      init.getIRVariable() instanceof IRThisVariable
+    )
+  )
+}
+
+/** Holds if `instr` flows to `succ`. */
+predicate successor(Instruction instr, Instruction succ) {
+  flowsToSink(succ) and
```

Comment:
I'm confused about the mutual recursion between `successor` and `flowsToSink`. I'll ping you on Slack so we can figure it out in a Zoom call.

Reply:
Your confusion was well-founded :) Fixed in 896dee5fb79c528400f082144d67bb76de27f273.